### PR TITLE
Improve definition of filterables fields

### PR DESCRIPTION
**Pull request by lorado**
Hi. Here is me again... 🤣 

in my project I wanted to define filterable fields type, but there was no option to do it. So I improved parsing of given values:

- There is new `TypedFilterInterface`, with method `getType()`, that should return wished input type. This type will be used, if instance of class that implements `TypedFilterInterface` defined in `YourDefinition::getFilterables()`
- there is still ability to define input type via array. So with array devinition it is possible to define type and use a closure (see examples below)
- all possible definitions are tested with unit tests

So here are all possible definitions of filterables:

```php
// NOTE: ComplexFilter implements TypedFilterInterface! 
public function getFilterable() {
	return [
		// Simple equality check (or "in" if value is an array). Type is Type::json()
		'id'       => new EqualsOrContainsFilter(),
		
		// Customized filter. Type is Type::json()
		"nameLike" => function($builder, $value) {
			return $builder->whereRaw('name like ?', $value);
		},
		
		// type is Type::string()
		"anotherFilter" => [
			"type" => Type::string(),
			"resolver" => function($builder, $value) {
				return $builder->whereRaw('anotherFilter like ?', $value);				
			}
		],
		
		// type is what is returned from `ComplexFilter::getType()`.
		// This is the preffered way to define filters, as it keeps definitions code clean
		"complexFilter" => new ComplexFilter(),
		
		// type in array will be overriden by what is returned from `ComplexFilter::getType()`.
		// this kind of difinition is not clear, but is implemented for backward compatibilities. Please don't use it
		"complexFilter2" => [
			"type" => Type::int(),
			"resolver" => new ComplexFilter()
		],
	];
}
```

I also updated README and listed there all possible usages (as above).

There is NO breaking changes. Old definitions are still working, with default `Type::json()` type.

As this is already 4th pull request, after merging some of PRs may require manually merge from me. So... let me know...

### CODE REVIEW

File: src/Filter/TypedFilterInterface.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Abstract filter
+ */
+interface TypedFilterInterface extends FilterInterface {
+
+	/**
+	 * Defines GraphQL Type if current filter
+	 * @return Type|InterfaceType|UnionType|ScalarType|InputObjectType|EnumType|ListOfType|NonNull $wrappedType
+	 */
+	public function getType();
```
[the remark this answers is not in the document]
Comment (lorado) [reply src/Filter/TypedFilterInterface.php]: I was not sure on this part. I think it is important here, that `getType()` don't return an `ObjectType`, as graphql can't resolve it then. Input types should be of type `InputObjectType`, or some of scalar type, or listOf, or nonNull. I see now, that there is `Type` in the beginning... I think it is wrong. Maybe I was a little too fast...
I will test it this week more deeper, and update this PR

File: src/Grammar/Grammar.php [hunk 2]
```diff
@@ -47,13 +47,25 @@ private function getBuilder(Builder $builder, $filter, $operator = "AND") {
...
+			$resolver = $filter['filter'];
+			// check, if we got an array, and try fetch resolver
+			if (is_array($filter['filter'])) {
+				if (key_exists('resolver', $filter['filter'])) {
+					$resolver = $filter['filter']['resolver'];
+				} else {
+					throw new FilterException("Invalid filter for $filter[key]");
+				}
+			}
+
```
[the remark this answers is not in the document]
Comment (lorado) [reply src/Grammar/Grammar.php]: Ok, gonna change it near end of the week

File: src/Filter/TypedFilterInterface.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (lorado) [reply src/Filter/TypedFilterInterface.php]: So I digged a little deeper, there is an `InputType` Interface in webonyx library, and that is what I was thinking/talking about. Now is it much cleaner ;)

File: src/Grammar/Grammar.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (lorado) [reply src/Grammar/Grammar.php]: done

**Comment by drasill** [comment]
Thank you !

**Pull request merged**